docs/Client.md

```diff
@@ -0,0 +1,989 @@
+
+Returns: `void | Promise<StreamData>` - Only returns a `Promise` if no `callback` argument was passed
+
+#### Parameter: `StreamFactoryData`
+
+* **statusCode** `number`
+* **headers** `http.IncomingHttpHeaders`
+* **opaque** `unknown`
+
+#### Parameter: `StreamData`
+
+* **opaque** `unknown`
+* **trailers** `Record<string, string>`
+
+#### Example 1 - Basic GET stream request
+
+```js
+'use strict'
+const { createServer } = require('http')
+const { Client } = require('undici')
+const { Writable } = require('stream')
+
+const server = createServer((request, response) => {
+  response.end('Hello, World!')
+})
+
+server.listen(() => {
+  const client = new Client(`http://localhost:${server.address().port}`)
+
+  const bufs = []
+
+  client.stream({
+    path: '/',
+    method: 'GET',
+    opaque: { bufs }
+  }, ({ statusCode, headers, opaque: { bufs } }) => {
+    console.log(`response received ${statusCode}`)
+    console.log('headers', headers)
+    return new Writable({
+      write (chunk, encoding, callback) {
+        bufs.push(chunk)
+        callback()
+      }
+    })
+  }).then(({ opaque: { bufs } }) => {
+    console.log(Buffer.concat(bufs).toString('utf-8'))
+
+    client.close()
+    server.close()
+  }).catch(error => {
+    console.error(error)
+  })
+})
+```
+
+#### Example 2 - Stream to Fastify Response
+
+In this example, a (fake) request is made to the fastify server using `fastify.inject()`. This request then executes the fastify route handler which makes a subsequent request to the raw Node.js http server using `undici.client.stream()`. The fastify response is passed to the `opaque` option so that undici can tap into the underlying writable stream using `response.raw`. This methodology demonstrates how one could use undici and fastify together to create fast-as-possible requests from one backend server to another.
+
+```js
+'use strict'
+
+const { createServer } = require('http')
+const undici = require('undici')
+const fastify = require('fastify')
+
+const nodeServer = createServer((request, response) => {
+  response.end('Hello, World! From Node.js HTTP Server')
+})
+
+const fastifyServer = fastify()
+
+nodeServer.listen(0, () => {
+  console.log('Node Server listening')
+
+  const client = new undici.Client(`http://localhost:${nodeServer.address().port}`)
+
+  fastifyServer.route({
+    url: '/',
+    method: 'GET',
+    handler: (request, response) => {
+      client.stream({
+        path: '/',
+        method: 'GET',
+        opaque: response
+      }, ({ opaque }) => opaque.raw)
+    }
+  })
+
+  fastifyServer.listen(0, () => {
+    console.log('Fastify Server listening')
+    fastifyServer.inject({
+      path: '/',
+      method: 'GET'
+    }).then(({ statusCode, body }) => {
+      console.log(`response received ${statusCode}`)
+      console.log('body:', body)
+
+      client.close()
+      fastifyServer.close()
+      nodeServer.close()
+    })
+  })
+})
```

Comment:
Here you are mixing callbacks and promises, which is generally not recommended.
Also, using `listen` and `inject` together is not very useful, how about using another undici instance for running the request?

Reply:
Yes unfortunately, node `server.listen()` does not return a promise. But I think I'll modify the fastify piece to be completely promise based + incorporate another undici instance for querying fastify. Good idea!